holoviews/element/comparison.py

```diff
@@ -648,3 +656,140 @@ def __init__(self, *args, **kwargs):
+
+    Specify a custom match_fn to control exactly what (and 
+    in what order) to compare.
+
+    By default, archive members will simply be opened and read for
+    comparison; specify alternatives for particular file types
+    by adding them to load_fns.
+    """
+
+    match_fn = param.Callable(default=match_by_name,doc="""
+        Function that matches up files in a 'reference' archive to
+        those in a 'test' archive.
+
+        Should accept two archive readers and return:
+
+            1. a dict-like mapping of files to compare (using a value 
+               of None to indicate missing test data);
+
+            2. a list (potentially empty) of files present in the test
+               that are not present in the reference.
+        """)
+
+    load_fns = param.Dict(default=defaultdict(lambda: read, {'.hvz': loadpickle}))
+
+    def __init__(self,ref,test,**params):
+        super(ArchiveComparison,self).__init__(**params)
+        self.ref = self._read_archive(ref)
+        self.test = self._read_archive(test)        
+
+
+    def equal(self):
+        """
+        Return True if the two archives contain the same contents,
+        otherwise False.
+        
+        Stops at first failure to match (subsequently available as
+        .first_failure).
+        """
+        for result in self.comparisons():
+            if not result.match:
+                self.first_failure = result
+                return False
+        return True
+
+
+    def comparisons(self):
+        """
+        Return a generator to allow stepping through comparisons.
+
+        Each comparison yields a ComparisonResult.
+        """
+        matches,extra = self.match_fn(self.ref,self.test)
+
+        for unused in extra:
+            yield self._result(unused,False,'Extra test member')
+        
+        for rmember in matches:
+            if matches[rmember] is None:
+                yield self._result(rmember,False,'Test is missing member')
+            else:
+                load_fn = self.load_fns[os.path.splitext(rmember)[1].lower()]
+                r = load_fn(self.ref.open_member(rmember))
+                try:
+                    t = load_fn(self.test.open_member(matches[rmember]))
+                except:
+                    yield self._result(rmember,False,'Cannot open test and ref members with same function (%s): %s'%(lf,traceback.format_exc()))
+                    continue # sorry
+                    
+                # ok finally we discovered we can actually do a comparison
+                try:
+                    Comparison.assertEqual(r,t)
+                except:
+                    yield self._result(rmember,False,traceback.format_exc())
+                else:
+                    yield self._result(rmember,True,None)
+                
+
+
+    def _read_archive(self,archive_location):
+        if not os.path.exists(archive_location):
+            raise IOError("Cannot find archive %s"%archive_location)
+
```

Comment:
I would stick the expected file extension (or more likely a list of file extensions) on the ``ArchiveReaders``. The I would just have a class attribute:

```python
class ArchiveComparison(param.Parameterized):
    readers = [Directory, ZipFile, TarFile]
```

Then ``_read_archive`` can iterate through to find an ``ArchiveReader`` it can use. As ``Directory`` isn't associated with a file extension (and you probably need to check a list of possible file extensions) I would probably check if the ``ArchiveReader`` can be used with an ``applicable`` method (or something like that returning a boolean).